File: src/rpf/runtime/transport.py
from __future__ import annotations

import select
import socket
from typing import Optional, Tuple


class UdpTransport:
    def __init__(self, bind_address: str, bind_port: int, recv_buf_size: int = 65535) -> None:
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._sock.bind((bind_address, bind_port))
        self._recv_buf_size = recv_buf_size

    def recv(self, timeout_s: float) -> Optional[Tuple[bytes, tuple[str, int]]]:
        readable, _, _ = select.select([self._sock], [], [], max(0.0, timeout_s))
        if not readable:
            return None
        data, addr = self._sock.recvfrom(self._recv_buf_size)
        return data, (str(addr[0]), int(addr[1]))

    def send(self, payload: bytes, address: str, port: int) -> None:
        self._sock.sendto(payload, (address, int(port)))

    def close(self) -> None:
        self._sock.close()

File: src/irp_rust/src/runtime/transport.rs
use std::io;
use std::net::{SocketAddr, UdpSocket};
use std::time::Duration;

pub struct UdpTransport {
    sock: UdpSocket,
    recv_buf_size: usize,
}

impl UdpTransport {
    pub fn bind(bind_address: &str, bind_port: u16, recv_buf_size: usize) -> io::Result<Self> {
        let sock = UdpSocket::bind((bind_address, bind_port))?;
        Ok(Self {
            sock,
            recv_buf_size,
        })
    }

    pub fn recv(&self, timeout: Duration) -> io::Result<Option<(Vec<u8>, SocketAddr)>> {
        self.sock.set_read_timeout(Some(timeout))?;
        let mut buf = vec![0_u8; self.recv_buf_size];
        match self.sock.recv_from(&mut buf) {
            Ok((n, addr)) => {
                buf.truncate(n);
                Ok(Some((buf, addr)))
            }
            Err(err)
                if err.kind() == io::ErrorKind::WouldBlock
                    || err.kind() == io::ErrorKind::TimedOut =>
            {
                Ok(None)
            }
            Err(err) => Err(err),
        }
    }

    pub fn send(&self, payload: &[u8], address: &str, port: u16) -> io::Result<usize> {
        self.sock.send_to(payload, (address, port))
    }
}
